noir-projects/aztec-nr/aztec/src/utils/bytes/bounded_vec.nr

```diff
@@ -0,0 +1,324 @@
+/// sure if they ever will be.
+
+/// Converts the input bytes into a BoundedVec of fields. A Field is ~254 bits meaning that each field can store 31
+/// whole bytes.
+///
+/// The input bytes are chunked into chunks of 31 bytes. Each 31-byte chunk is viewed as big-endian, and is converted
+/// into a Field. If the last chunk is less than 31 bytes long, then only the relevant bytes are considered.
+/// For example, [1, 10, 3] is encoded as [1 * 256^2 + 10 * 256 + 3]
+/// Note: ceil(N / 31) = (N + 30) / 31
+pub fn be_bytes_31_to_fields<let N: u32>(
+    bytes: BoundedVec<u8, N>,
+) -> BoundedVec<Field, (N + 30) / 31> {
+    let mut fields = BoundedVec::new();
+
+    // There are "bytes.len() / 31" whole fields that can be populated.
+    for i in 0..bytes.len() / 31 {
+        let mut field = 0;
+        for j in 0..31 {
+            // Shift the existing value left by 8 bits and add the new byte
+            field = field * 256 + bytes.get(i * 31 + j) as Field;
+        }
+        fields.push(field);
+    }
+
+    // Note: if 31 divides bytes.len(), then this loop does not execute.
+    // ceil(bytes.len()/31) - floor(bytes.len()/31) = 1, unless 31 divides bytes.len(), in which case it's 0.
+    for _ in 0..((bytes.len() + 30) / 31) - (bytes.len() / 31) {
+        let mut final_field = 0;
+        let final_field_idx = fields.len();
+        let final_offset = final_field_idx * 31;
+        // bytes.len() - ((bytes.len() / 31) * 31) = bytes.len() - floor(bytes.len()/31) * 31 = the number of bytes
+        // to go in the final field.
+        for j in 0..bytes.len() - ((bytes.len() / 31) * 31) {
+            // Shift the existing value left by 8 bits and add the new byte
+            final_field = final_field * 256 + bytes.get(final_offset + j) as Field;
+        }
+
+        fields.push(final_field);
+    }
+
+    fields
+}
+
+/// Converts an input BoundedVec of fields into a BoundedVec of bytes in big-endian order.
+///
+/// Each input field must contain at most 31 bytes (this is constrained to be so). Note that the last field might
+/// contain less than 31 bytes. This is communicated through the M parameter.
+///
+/// The M parameter is flexible rather than computed from N because this function is commonly used to convert fields
+/// back to bytes after a previous bytes->fields conversion. Since the original byte length may not be a multiple
+/// of 31, having a flexible M parameter allows us to output the same number of bytes as the original input.
+///
+/// Each field (other than the last which might be shorter) is converted into 31 big-endian bytes, and the
+/// resulting 31-byte chunks are concatenated back together in the order of the original fields. You'll end up
+/// with quite a strange ordering of bytes if you use this particular function.
+pub fn fields_to_be_bytes_31<let N: u32, let M: u32>(
+    fields: BoundedVec<Field, N>,
+) -> BoundedVec<u8, M> {
```

Comment:
This function is quite different from the array one because we need to perform different checks on the last field based on the input vec being full or not (see comment on line 69).